storage/innobase/xtrabackup/src/xbstream.cc

```diff
@@ -668,31 +726,27 @@ static int mode_extract(int n_threads, int argc __attribute__((unused)),
       ds_set_pipe(ds_decrypt_zstd_ctxt, ds_decompress_zstd_ctxt);
     }
   }
-
-  stream = xb_stream_read_new();
-  if (stream == NULL) {
-    msg("%s: xb_stream_read_new() failed.\n", my_progname);
-    pthread_mutex_destroy(&mutex);
-    ret = 1;
-    goto exit;
+  data_threads = (extract_ctxt_t *)my_malloc(
+      PSI_NOT_INSTRUMENTED, sizeof(extract_ctxt_t) * (n_threads + 1),
+      MYF(MY_FAE));
+  for (int i = 0; i < n_threads; i++) {
+    data_threads[i].thread_id = i;
+    data_threads[i].filehash = filehash;
+    data_threads[i].ds_ctxt = ds_ctxt;
+    data_threads[i].ds_decompress_quicklz_ctxt = ds_decompress_quicklz_ctxt;
+    data_threads[i].ds_decompress_lz4_ctxt = ds_decompress_lz4_ctxt;
+    data_threads[i].ds_decrypt_uncompressed_ctxt = ds_decrypt_uncompressed_ctxt;
+    data_threads[i].ds_decrypt_quicklz_ctxt = ds_decrypt_quicklz_ctxt;
+    data_threads[i].ds_decrypt_lz4_ctxt = ds_decrypt_lz4_ctxt;
+    data_threads[i].mutex = &mutex;
   }
 
-  ctxt.stream = stream;
-  ctxt.ds_ctxt = ds_ctxt;
-  ctxt.ds_decompress_quicklz_ctxt = ds_decompress_quicklz_ctxt;
-  ctxt.ds_decompress_lz4_ctxt = ds_decompress_lz4_ctxt;
-  ctxt.ds_decompress_zstd_ctxt = ds_decompress_zstd_ctxt;
-  ctxt.ds_decrypt_uncompressed_ctxt = ds_decrypt_uncompressed_ctxt;
-  ctxt.ds_decrypt_quicklz_ctxt = ds_decrypt_quicklz_ctxt;
-  ctxt.ds_decrypt_lz4_ctxt = ds_decrypt_lz4_ctxt;
-  ctxt.ds_decrypt_zstd_ctxt = ds_decrypt_zstd_ctxt;
-  ctxt.mutex = &mutex;
-
-  tids = new pthread_t[n_threads];
-  retvals = new void *[n_threads];
+  tids = new pthread_t[n_threads + 1];
+  retvals = new void *[n_threads + 1];
 
   for (i = 0; i < n_threads; i++)
-    pthread_create(tids + i, NULL, extract_worker_thread_func, &ctxt);
+    pthread_create(tids + i, NULL, extract_worker_thread_func,
```

Comment:
handle error of pthread_create

Reply:
Done. `pthread_cancel` will signal already started threads to cleanup and exit.